In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import pickle
from pandas_profiling import ProfileReport

In [2]:
df = pd.read_csv("/home/meta-sapien/ineuron/ML/linear Regression/Advertising.csv")
df.head(5)

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

Few Questions: 
1. Build a model that can predict sales
2. Which mode of advertisement will be more effective if i invest in TV or Radio or Newspaper?
3. If relations are possible?

In [3]:
df.describe()

Unnamed: 0          TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   100.500000  147.042500   23.264000   30.554000   14.022500
std     57.879185   85.854236   14.846809   21.778621    5.217457
min      1.000000    0.700000    0.000000    0.300000    1.600000
25%     50.750000   74.375000    9.975000   12.750000   10.375000
50%    100.500000  149.750000   22.900000   25.750000   12.900000
75%    150.250000  218.825000   36.525000   45.100000   17.400000
max    200.000000  296.400000   49.600000  114.000000   27.000000

In [4]:
pf = ProfileReport(df)

In [5]:
pf.to_widgets()

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]

Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render widgets:   0%|          | 0/1 [00:00<?, ?it/s]

In [6]:
pf.to_file("test.html")

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]

# Building the Model

1. TV Vs Sales:

In [7]:
x = df[["TV"]]
y = df.Sales

y = m.x + c

In [8]:
from sklearn.linear_model import LinearRegression

In [9]:
linear = LinearRegression()
linear.fit(x,y)

LinearRegression()

In [10]:
linear.intercept_ #c

7.032593549127693

In [11]:
linear.coef_ #m

array([0.04753664])

y = 0.04753.x + 7.0325

In [12]:
file = "linear_reg.sav"
pickle.dump(linear,open(file,'wb'))

In [13]:
linear.predict([[45]])

array([9.17174237])

if we are spending 45 on TV then sales will be 9.17,

In [14]:
l = [10,20,30,40,100,120]

In [15]:
for i in l:
    print(linear.predict([[i]]))

[7.50795995]
[7.98332636]
[8.45869276]
[8.93405917]
[11.78625759]
[12.7369904]


#Note: Passing the 2D inputs

In [16]:
saved_model = pickle.load(open(file,'rb'))
saved_model.predict([[20]])

array([7.98332636])

In [17]:
linear.score(x,y) #calculates the accuracy of model. Not the best way since we passed the same data we gave during learing::

0.611875050850071